### ISSUE TITLE
Possible support for weighted average loss calculation in FusedLinearCrossEntropy kernel

### ISSUE THREAD

**Issue by ChenlongDeng** [linkedin/Liger-Kernel#338]
### 🚀 The feature, motivation and pitch

Hi, great job on the FusedLinearCrossEntropy kernel! I’ve found it very efficient for model training. However, it seems to lack support for custom token-wise weighted loss calculation. For instance, users might input a 2D tensor containing individual loss weights for each token and calculate a weighted average loss based on those values. Does this version support that functionality, or is it something that might be added in the future?

**Comment by shivam2199** [comment]
Hi @Tcc0403 @vaibhavjindal  — I'd like to pick this up if it's still wanted.

Quick clarification first: PR #420 (closing #404) added ce_weight for per-class weighting, which matches torch.nn.CrossEntropyLoss(weight=...). This issue (#338) asks for per-token weighting — a distinct feature that still isn't supported, so I don't think #420 fully covered it. Let me know if I've misread.

Proposed API
Add a new optional arg to LigerFusedLinearCrossEntropyLoss.forward:


`def forward(self, lin_weight, _input, target, bias=None, token_weights=None):`

`token_weights`: float tensor of shape `(BT,)` (or `(B, T) flattened)`, non-negative, matching target's leading dim. Applied multiplicatively to each token's loss and its contribution to the gradient.
Stacks cleanly with existing `ce_weight` (multiplication: `loss_i *= ce_weight[target_i] * token_weights[i])`.

Design questions before I start:

- Normalization under reduction='mean' — match PyTorch's weighted CE semantics (divide by sum(token_weights[target_mask] * ce_weight[target[mask]])), or simpler sum(token_weights[target_mask])? I'd match PyTorch.
- Interaction with use_token_scaling — both are per-token multiplicative factors. Allow composition (user gets both multiplied), or raise if both set?
- Implementation: post-hoc scalar mul in the chunk loop (like current use_token_scaling) is simpler and enough for correctness. Integrating the weight inside liger_cross_entropy_kernel would be faster by avoiding an extra (BT,) read per chunk. Any preference?
- ignore_index — I'd force loss = 0 for ignored tokens regardless of their weight value (matches use_token_scaling behavior). OK?
If the direction looks right, I'll open a PR with op changes, module wrapper, and tests (parity against `F.cross_entropy(reduction='none') * weights, plus gradient check)`. Happy to adjust scope based on your preferences.
